### DISCUSSION TITLE
Flutter sqflite multi tables

### DISCUSSION THREAD

**Discussion by Jef2403** [community/community#63150] [Programming Help] [answered]
### Body

Bonjour,
J'ai commencé à développer une application Flutter / Sqflite avec 1 base de données composée de 3 tables : 2 tables de référence (A et B)et 1 table de jonction / relation (C) ["create table A(id_A integer primary key autoincrement not null, nom_A text(4) not null unique);" "create table B(id_B integer primary key autoincrement not null, nom_B text(30) not null unique);" "create table C(id_C integer primary key autoincrement not null, id_C_A integer not null, id_C_B integer not null, foreign key (id_C_A) references A(id_A), foreign key (id_C_B) references B(id_B));". Avez vous des exemples à me conseiller ? Pour l'instant j'arrive a bien créer mes 2 tables (A et B) et les mettre à jour (CRUD), mais j'ai un problème pour le code de la table de relation (C) ! je voudrais dans la table C choisir un nom de la table A et l'associer à un nom de la table B. Je vous remercie par avance de votre aide.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Jef2403** [comment]
Hello
I started to develop a Flutter / Sqflite application with 1 database consisting of 3 tables:  2 reference tables (A and B) and 1 junction/relationship table (C)  ["create table A(id_A integer primary key autoincrement not null, nom_A text(4) not null unique);" "create table B(id_B integer primary key autoincrement not null, nom_B text(30) not null unique);" "create table C(id_C integer primary key autoincrement not null, id_C_A integer not null, id_C_B integer not null, foreign key (id_C_A) references A(id_A), foreign key (id_C_B) references B(id_B));" Do you have any examples to advise me?  For now I manage to create my 2 tables (A and B) and update them (CRUD),   but I have a problem with the code of the relationship table (C)! I would like to choose a name from table A in table C and associate it with a name from table B. Thank you in advance for your help.

**Comment by Jef2403** [comment]
![Screen_add_Tab_A](https://github.com/community/community/assets/134713432/ce6ddcfb-8926-44af-a461-1d58d999305a)
![Screen_add_Tab_B](https://github.com/community/community/assets/134713432/46f63ae2-0d8f-45be-a25a-70a87e67bb81)
![Screen_add_Tab_C](https://github.com/community/community/assets/134713432/b82d96bd-6591-4e9b-b6f8-d750873a0e95)
![Screen_list_Tab_A](https://github.com/community/community/assets/134713432/b5a1df17-8f04-464b-99f8-7adb7425f168)
![Screen_list_Tab_B](https://github.com/community/community/assets/134713432/d8bec448-2d6b-4775-998c-8714fc2e88db)
Please find enclosed the screenshots
Thanks

**Comment by RamtinKosari** [comment]
Bonjour!

Il semble que vous ayez déjà créé avec succès vos tables de référence A et B, ainsi que votre table de jonction C dans votre application Flutter utilisant Sqflite. Maintenant, vous souhaitez associer un nom de la table A à un nom de la table B dans la table C. Je vais vous fournir un exemple simple de code pour vous aider à comprendre comment réaliser cela.

Supposons que vous ayez des classes de modèle pour vos tables A, B et C. Voici un exemple de code pour la classe C :
```c++
class C {
  int id;
  int id_A;
  int id_B;
  
  C({this.id, this.id_A, this.id_B});
  
  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'id_A': id_A,
      'id_B': id_B,
    };
  }
  
  factory C.fromMap(Map<String, dynamic> map) {
    return C(
      id: map['id'],
      id_A: map['id_A'],
      id_B: map['id_B'],
    );
  }
}
```
Maintenant, pour associer les noms des tables A et B dans la table C, vous pouvez ajouter des propriétés supplémentaires à la classe C :
```c++
class C {
  int id;
  int id_A;
  int id_B;
  String nom_A; // Nom de la table A
  String nom_B; // Nom de la table B
  
  C({this.id, this.id_A, this.id_B, this.nom_A, this.nom_B});
  
  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'id_A': id_A,
      'id_B': id_B,
      'nom_A': nom_A,
      'nom_B': nom_B,
    };
  }
  
  factory C.fromMap(Map<String, dynamic> map) {
    return C(
      id: map['id'],
      id_A: map['id_A'],
      id_B: map['id_B'],
      nom_A: map['nom_A'],
      nom_B: map['nom_B'],
    );
  }
}
```
Maintenant, lorsque vous effectuez des opérations CRUD sur la table C, vous pouvez également associer les noms des tables A et B. Par exemple, lors de l'insertion d'une nouvelle entrée dans la table C :
```c++
Future<void> insertC(C c) async {
  final db = await database;
  await db.insert('C', c.toMap());
}
```
Avec cette approche, vous pouvez associer les noms des tables A et B à chaque entrée de la table C et les récupérer lorsque vous lisez les données de la table C.

**Reply by Jef2403** [reply]
Merci beaucoup je vais faire les tests

**Comment by Jef2403** [comment]
Bonjour,
Merci beaucoup Ramtin de ces informations !
Je vais appliquer vos codes et je vous tiendrai au courant.
J'ai un autre problème dans l'écran de saisie dans la table C. J'ai mis 2
combobox, l'une pour la table A et l'autre pour la table B, je n'arrive pas
à récupérer les données dans sqflite !
Avec des listes statistiques (locales) ça marche très bien. C'est vrai que
je suis débutant dans Flutter !
Encore merci de votre aide

Le jeu. 24 août 2023 à 09:17, Ramtin Kosari ***@***.***> a
écrit :

**Reply by RamtinKosari** [reply]
Il semble que vous rencontriez des difficultés pour récupérer les données des tables A et B à partir de la table C dans sqflite en utilisant deux combobox. Vous mentionnez que cela fonctionne bien avec des listes statiques locales, mais vous avez du mal à le faire fonctionner avec sqflite. Comme vous êtes débutant dans Flutter, je vais vous donner quelques conseils pour résoudre ce problème.

Voici quelques étapes à suivre pour récupérer les données des tables A et B dans la table C à l'aide de sqflite :

1. Assurez-vous que vous avez correctement configuré et initialisé votre base de données sqflite. Vérifiez que vous avez créé les tables A, B et C avec les bonnes colonnes et types de données.

2. Assurez-vous d'avoir les bons imports pour utiliser sqflite dans votre fichier Flutter. Vous devriez avoir import 'package:sqflite/sqflite.dart'; et import 'package:path/path.dart'; pour utiliser les fonctionnalités de sqflite.

3. Vérifiez que vous avez correctement ouvert la connexion à votre base de données sqflite en utilisant await openDatabase(path), où path est le chemin de votre base de données.

4. Pour récupérer les données des tables A et B, vous devez effectuer des requêtes SQL appropriées à l'aide de sqflite. Par exemple, pour récupérer toutes les données de la table A, vous pouvez utiliser la requête suivante :
```dart
List<Map<String, dynamic>> tableAData = await db.query('tableA');
```
Assurez-vous d'adapter cette requête à votre structure de table et de récupérer les colonnes nécessaires.

5. De même, pour récupérer les données de la table B, vous pouvez utiliser une requête similaire :
```dart
List<Map<String, dynamic>> tableBData = await db.query('tableB');
```
6. Une fois que vous avez récupéré les données des tables A et B, vous pouvez les utiliser dans vos combobox pour afficher les options à l'utilisateur. Assurez-vous d'extraire les valeurs appropriées à partir des résultats des requêtes et de les utiliser pour remplir les combobox.

7. Lorsque l'utilisateur sélectionne une option dans les combobox, vous pouvez récupérer la valeur sélectionnée et effectuer les opérations nécessaires dans votre table C, en utilisant les données récupérées précédemment.
Assurez-vous de gérer les erreurs et les exceptions correctement lors de l'utilisation de sqflite. Vous pouvez également consulter la documentation officielle de sqflite pour plus d'exemples et de détails sur son utilisation dans Flutter.

J'espère que ces conseils vous aideront à résoudre votre problème avec sqflite. N'hésitez pas à poser des questions supplémentaires si vous avez besoin d'aide supplémentaire. Bonne chance avec votre développement Flutter !

**Reply by RamtinKosari** [reply, marked as answer]
Je serai heureux et heureux si vous marquez cette réponse comme "répondu", merci beaucoup ❤️

**Reply by Jef2403** [reply]
Super ! Très complet vos explications ! je vais les appliquer et je vous tiens au courant.

**Reply by RamtinKosari** [reply]
merci beaucoup

**Reply by Jef2403** [reply]
Ramtin je vous ai rajouté dans mon projet "tablesabc" dans Github

**Comment by Jef2403** [comment]
Bonjour Ramtin, je n'arrive toujours pas à récupérer les données sqflite dans la combobox (ex : ComboBoxA) ! voici mon code :
//Ecran ajout des Tab_C
//import "dart:io";
import 'package:flutter/material.dart';
import 'package:sqflite_common_ffi/sqflite_ffi.dart';
//import 'package:tablesabc/screens/viewtab_A.dart';
//import 'package:tablesabc/db_helper/repository.dart';
import 'package:tablesabc/db_helper/database_connection.dart';
//import 'package:tablesabc/models/tab_A.dart';
import 'package:tablesabc/services/tab_Aservice.dart';

const List<String> list1 = <String>['One', 'Two', 'Three', 'Four'];
//
var db = openDatabase('db_tablesabc');

final _tab_AService = Tab_AService();
//List<Map<String, dynamic>> listeA = _tab_AService.readAllTab_A();
List<Map<String, dynamic>> listeA = db.query('SELECT nom_A FROM tab_A');
//List<Map<String, dynamic>> listeA = db._queryDB();

const List<String> list2 = <String>['Pomme', 'Banane', 'Cerise', 'Ananas'];
//modif

// = await db.rawQuery('SELECT * FROM tab_A');

class AddTab_C extends StatefulWidget {
  const AddTab_C({Key? key, required tab_C}) : super(key: key);

  @override
  State<AddTab_C> createState() => _AddTab_CState();
}

class _AddTab_CState extends State<AddTab_C> {
  String dropdownValue1 = list1.first;
  String dropdownValue2 = list2.first;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Tables ABC')),
      body: SingleChildScrollView(
        child: Container(
          padding: const EdgeInsets.all(16.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              const Text(
                'Ajout nouveau Tab_C',
                style: TextStyle(
                    fontSize: 20,
                    color: Colors.teal,
                    fontWeight: FontWeight.w500),
              ),
              const SizedBox(
                height: 20.0,
              ),
              Row(
                children: [
                  TextButton(
                      style: TextButton.styleFrom(
                          foregroundColor: Colors.white,
                          backgroundColor: Colors.teal,
                          textStyle: const TextStyle(fontSize: 15)),
                      onPressed: () async {},
                      child: const Text('Sauvegarde Details')),
                  const SizedBox(
                    width: 10.0,
                  ),
                  TextButton(
                      style: TextButton.styleFrom(
                          foregroundColor: Colors.white,
                          backgroundColor: Colors.red,
                          textStyle: const TextStyle(fontSize: 15)),
                      onPressed: () => dropdownValue1.isEmpty,
                      child: const Text('Clear Details'))
                ],
              ),

//Appel ComboBox1
              //const DropdownMenu1(),
              const DropdownMenu1(),

//Appel ComboBox2
              const DropdownMenu2(),
//Appel ComboBoxA
              const DropdownMenuA(),
            ],
          ),
        ),
      ),
    );
  }
}

//ComboBox1
class DropdownMenu1 extends StatefulWidget {
  const DropdownMenu1({super.key});
//Modif

  @override
  State<DropdownMenu1> createState() => _DropdownMenu1State();
}

class _DropdownMenu1State extends State<DropdownMenu1> {
  String dropdownValue1 = list1.first;

  @override
  Widget build(BuildContext context) {
    return DropdownMenu<String>(
      initialSelection: list1.first,
      onSelected: (String? value) {
        // This is called when the user selects an item.
        setState(() {
          dropdownValue1 = value!;
        });
      },
      dropdownMenuEntries: list1.map<DropdownMenuEntry<String>>((String value) {
        return DropdownMenuEntry<String>(value: value, label: value);
      }).toList(),
    );
  }
}

//ComboBox2
class DropdownMenu2 extends StatefulWidget {
  const DropdownMenu2({super.key});

  @override
  State<DropdownMenu2> createState() => _DropdownMenu2State();
}

class _DropdownMenu2State extends State<DropdownMenu2> {
  String dropdownValue2 = list2.first;

  @override
  Widget build(BuildContext context) {
    return DropdownMenu<String>(
      initialSelection: list2.first,
      onSelected: (String? value) {
        // This is called when the user selects an item.
        setState(() {
          dropdownValue2 = value!;
        });
      },
      dropdownMenuEntries: list2.map<DropdownMenuEntry<String>>((String value) {
        return DropdownMenuEntry<String>(value: value, label: value);
      }).toList(),
    );
  }
}

//ComboBoxA
class DropdownMenuA extends StatefulWidget {
  const DropdownMenuA({super.key});

  @override
  State<DropdownMenuA> createState() => _DropdownMenuAState();
}

class _DropdownMenuAState extends State<DropdownMenuA> {
  //List<Map<String, dynamic>> listeA = await db.query('tab_A');

  String dropdownValueA = listA.first;

  @override
  Widget build(BuildContext context) {
    return DropdownMenu<String>(
      initialSelection: listA.first,
      onSelected: (String? value) {
        // This is called when the user selects an item.
        setState(() {
          dropdownValueA = value!;
        });
      },
      dropdownMenuEntries: listA.map<DropdownMenuEntry<String>>((String value) {
        return DropdownMenuEntry<String>(value: value, label: value);
      }).toList(),
    );
  }
}
**********************************
Merci par avance de votre aide
